### PR TITLE
[Merged by Bors] - TY-2805 Do not expose paging info of ActiveSearch to API.

### PR DESCRIPTION
**Pull request by rustonaut**
This is archived to have an `ActiveSearch` in the API and another internal one.

Also change the internal `ActiveSearch` to no longer use `freezed` due to the shenanigans needed to be able to add additional methods on a `freezed` object (and it's  not needed anymore).

**References:**

- [TY-2805](https://xainag.atlassian.net/browse/TY-2805)

### CODE REVIEW

File: discovery_engine/lib/src/domain/models/active_search.dart [hunk 1]
```diff
@@ -12,28 +12,46 @@
...
+  final SearchBy searchBy;
+
+  const ActiveSearch({
+    required this.searchTerm,
+    required this.requestedPageNb,
+    required this.pageSize,
+    required this.searchBy,
+  });
+
+  ActiveSearch nextPageSearch() => ActiveSearch(
```
Comment (janpetschexain) [inline discovery_engine/lib/src/domain/models/active_search.dart:43]: ❓ is it required (by hive or for other reasons) to create a new object? otherwise we could just mutate the current object by incrementing the `requestedPageNb`. this would be more in line with the updating character of the function as it is used in the search manager.

Comment (janpetschexain) [inline discovery_engine/lib/src/domain/models/active_search.dart:43]: :art: the naming seems a bit redundant
```suggestion
  ActiveSearch nextPage() => ActiveSearch(
```

File: discovery_engine/lib/src/domain/models/active_search.dart [hunk 2]
```diff
@@ -12,28 +12,46 @@
...
+        searchTerm: searchTerm,
+        requestedPageNb: requestedPageNb + 1,
+        pageSize: pageSize,
+        searchBy: searchBy,
+      );
+
+  api.ActiveSearch toApiRepr() =>
```
Comment (acrrd) [inline discovery_engine/lib/src/domain/models/active_search.dart:50]: This object should be completely agnostic of the api models we are currently using

File: discovery_engine/lib/src/domain/models/active_search.dart (hunk 1 above)
Comment (acrrd) [reply discovery_engine/lib/src/domain/models/active_search.dart:43]: Also this data is a basically a POD, it should not contain any logic about what the next page is.

Comment (rustonaut) [reply discovery_engine/lib/src/domain/models/active_search.dart:43]: > question is it required (by hive or for other reasons) to create a new object? 

No, just less likely to cause problems due to accidental shared references in a GCed language.
And at least new/before the fields had been final.

> Also this data is a basically a POD, it should not contain any logic about what the next page is.

This is roughly the same design as `Document` which also has small helpers (in form of computed fields/getters).

Before we used `foo.copyWith(requestedPageNb: foo.requestedPageNb + 1)` ad-hoc, but that's a bit prone to mistakes.

Furthermore given that this type already contains the concept of paging it's not like it's some logic specific to some implementation (or at least not more then the fields already are).

If you have a better place to put the method I will move it there. I mean I did consider an extension trait but it's just not worth the additional overhead IMHO.

File: discovery_engine/lib/src/domain/models/active_search.dart (hunk 2 above)
Comment (rustonaut) [reply discovery_engine/lib/src/domain/models/active_search.dart:50]: It's the same approach as currently used by `Document`, it that wouldn't already have been this way in `Document` I would have placed it into `api/models/active_search.dart`. 

When doing so should I also move to `toApiDocument` method on `Document` from `domain/models` to `api/models`?

Comment (rustonaut) [reply discovery_engine/lib/src/domain/models/active_search.dart:50]: Through it also add some complexity  as you have to either write `ActiveSearch.fromInternalRepr(search)` instead of `search.toApiRepr()` or have to write a extension trait you then also have to import.

Comment (rustonaut) [reply discovery_engine/lib/src/domain/models/active_search.dart]: It's now a extension method.

File: discovery_engine/lib/src/domain/search_manager.dart [hunk 3]
```diff
@@ -121,23 +123,26 @@ class SearchManager {
...
+    final search = await _searchRepo.getCurrent();
 
     if (search == null) {
       const reason = SearchFailureReason.noActiveSearch;
       return const EngineEvent.nextSearchBatchRequestFailed(reason);
     }
 
     // lets update active search params
-    search = search.copyWith(requestedPageNb: search.requestedPageNb + 1);
+    search.requestedPageNb += 1;
```
Comment (janpetschexain) [inline discovery_engine/lib/src/domain/search_manager.dart:139]: 🔧 if we change the field directly, then we could remove the unused `ActiveSearch.nextPage()` method completely. is there a specific reason why the method is not used anymore (independent of whether the method mutates the object or creates a new object)?

❓ we probably can't mark the fields of `ActiveSearch` as private because of `hive`, right? otherwise we could better contain the changes to the object.

File: discovery_engine/lib/src/domain/models/document.dart [hunk 4]
```diff
@@ -69,6 +69,8 @@ class Document {
     this.isSearched = false,
   }) : timestamp = DateTime.now().toUtc();
 
+  //FIXME move this conversion to `api/models/document.dart` e.g. as an
```
Comment (janpetschexain) [inline discovery_engine/lib/src/domain/models/document.dart:72]: ❓ i wouldn't mind if this was done already in this pr, but we can also leave it for another pr. in the latter case, please add a ticket to the backlog.

File: discovery_engine/lib/src/domain/search_manager.dart (hunk 3 above)
Comment (rustonaut) [reply discovery_engine/lib/src/domain/search_manager.dart:139]: > we probably can't mark the fields [...] private [...]

Also we use the fields in various places outside of the module so we would need to add getters and and the `nextPage` method again which conflicts with the feedback from @acrrd :

> Also this data is a basically a POD,  it should not contain any logic about what the next page is.

It's also not in line with the other types in `(domain|api)/models/` which are all PODs.

Comment (rustonaut) [reply discovery_engine/lib/src/domain/search_manager.dart:139]: >  [..] could remove [..]

done

File: discovery_engine/lib/src/domain/models/document.dart (hunk 4 above)
Comment (rustonaut) [reply discovery_engine/lib/src/domain/models/document.dart]: Ok did it.

**Review by janpetschexain — APPROVED** [review]
lgtm

**Comment by rustonaut** [comment]
bors merge
